lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.Common.cpp

```diff
@@ -0,0 +1,275 @@
+
+#include "RuntimePlatformAgnosticPch.h"
+#include "UnicodeText.h"
+
+namespace PlatformAgnostic
+{
+namespace UnicodeText
+{
+CharacterTypeFlags GetLegacyCharacterTypeFlags(char16 ch)
+{
+    Assert(ch >= 128);
+
+    const char16 lineSeperatorChar = 0x2028;
+    const char16 paraSeperatorChar = 0x2029;
+
+    if (ch == lineSeperatorChar || ch == paraSeperatorChar)
+    {
+        return LineCharGroup;
+    }
+
+    CharacterClassificationType charClass = GetLegacyCharacterClassificationType(ch);
+
+    if (charClass == CharacterClassificationType::Letter)
+    {
+        return LetterCharGroup;
+    }
+    else if (charClass == CharacterClassificationType::DigitOrPunct)
+    {
+        return IdChar;
+    }
+    else if (charClass == CharacterClassificationType::Whitespace)
+    {
+        return SpaceChar;
+    }
+
+    return UnknownChar;
+}
+
+#if !defined(_WIN32) || ENABLE_TEST_PLATFORM_AGNOSTIC
+namespace Internal
+{
+template <typename CharType>
+inline bool isDigit(__in CharType c)
+{
+    return c >= '0' && c <= '9';
+}
+
+template <typename CharType>
+inline int readNumber(__inout CharType* &str)
+{
+    int num = 0;
+
+    // Count digits on each string
+    while (isDigit(*str))
+    {
+        num = (num * 10) + (*str - '0');
+        str++;
+    }
+
+    return num;
+}
+
+///
+/// Implementation of CompareString(NORM_IGNORECASE | SORT_DIGITSASNUMBERS)
+/// This code is in the common library so that we can unit test it on Windows too
+///
+/// Basic algorithm is as follows:
+/// 
+/// Iterate through both strings
+///  If either current character is not a number, compare the rest of the strings lexically
+///  else if they're both numbers:
+///     Count the leading zeroes in both strings and skip over them
+///     Read the two numbers
+///     If they're not the same, return accordingly
+///     If they are the same, but the leading zeroes are different, return whichever has fewer zeroes as "larger"
+///     Otherwise, they're the same to continue scanning
+///  else they're both not numbers:
+///     Compare lexically till we find a number
+///
+template <typename CharType>
+int LogicalStringCompareImpl(__in const CharType* p1, __in const CharType* p2)
+{
+    Assert(p1 != nullptr);
+    Assert(p2 != nullptr);
+
+    while (*p1 && *p2)
+    {
+        bool isDigit1 = isDigit(*p1);
+        bool isDigit2 = isDigit(*p2);
+
+        // If either character is not a number, compare lexically
+        if (isDigit1 ^ isDigit2)
+        {
+            while (*p1 && *p2)
+            {
+                CharType c1 = *p1;
+                CharType c2 = *p2;
+
+                if (c1 == c2)
+                {
+                    p1++; p2++;
+                    continue;
+                }
+
+                if (c1 < c2)
+                {
+                    return -1;
+                }
+                else
+                {
+                    return 1;
+                }
+            }
+
+            return 0;
```

Comment:
Actually this branch looks weird. If `isDigit(c1) != isDigit(c2)`, then `c1 != c2`. This doesn't need a `while` loop and only needs to compare c1 with c2 and return.